### PR TITLE
Biomass transport

### PR DESCRIPTION
**Pull request by lisazeyen**
- add option to split biomass potential to single nodes instead of global EU node
- transport of solid biomass between nodes (costs based on [JRC-EU-Times Bioenergy report](http://dx.doi.org/10.2790/01017) and depend on country and distance)
- add costs for upgrading biogas to gas
- smaller changes: 
-> drop nan values when removing industry electricity demand to avoid warning (runtime error)
-> add color for 'industry electricity' to color map in config.yaml
-> move input data for transport to Snakefile

### CODE REVIEW

File: scripts/prepare_sector_network.py [hunk 1]
```diff
@@ -1856,6 +2014,9 @@ def remove_h2_network(n):
     if "noH2network" in opts:
         remove_h2_network(n)
 
+    if not options["biomass_transport"]:
+        remove_biomass_transport(n)
+
```
Comment (fneum) [inline scripts/prepare_sector_network.py:2019]: Hmm, it's a bit odd to add biomass transport first and then remove it later if `biomass_transport: False` (same for H2 network above actually). Could we do something like:

```suggestion
    if options["biomass_transport"]:
        add_biomass_transport(n)
    else:
        add_biomass_copperplated(n)
```
where `add_biomass_copperplated(n)` is the current `add_biomass(n)` and `add_biomass_transport(n) includes the changes you made in this PR?

File: scripts/prepare_sector_network.py [hunk 2]
```diff
@@ -1375,22 +1412,51 @@ def add_biomass(network):
...
+                 e_initial=biomass_pot_node["solid biomass"].values)
 
     network.madd("Link",
                  ["biogas to gas"],
                  bus0="EU biogas",
                  bus1="EU gas",
                  bus2="co2 atmosphere",
                  carrier="biogas to gas",
+                 capital_cost=costs.loc["biogas upgrading", "fixed"],
+                 marginal_cost=costs.loc["biogas upgrading", "VOM"],
```
[the remark this answers is not in the document]
Comment (fneum) [reply scripts/prepare_sector_network.py:1429]: Ah, just saw they are already in 0.2.0.

[the remark this answers is not in the document]
Comment (fneum) [reply scripts/prepare_sector_network.py:1429]: Pulled this in via #126

File: scripts/prepare_sector_network.py [hunk 3]
```diff
@@ -51,6 +51,33 @@ def add_lifetime_wind_solar(n):
...
+    positive_order = candidates.bus0 < candidates.bus1
+    candidates_p = candidates[positive_order]
+    candidates_n = (candidates[~ positive_order]
+                    .rename(columns={"bus0": "bus1", "bus1": "bus0"}))
+    candidates = pd.concat((candidates_p, candidates_n), sort=False)
+
+    topo = candidates.groupby(["bus0", "bus1"], as_index=False).mean()
+    topo.rename(index=lambda x: prefix + topo.at[x, "bus0"]
+                + connector + topo.at[x, "bus1"],
+                inplace=True)
```
Comment (fneum) [inline scripts/prepare_sector_network.py:77]: Slightly more compact:
```suggestion
    topo.index = topo.apply(lambda x: prefix + x.bus0 + connector + x.bus1, axis=1)
```

File: scripts/prepare_sector_network.py [hunk 4]
```diff
@@ -51,6 +51,33 @@ def add_lifetime_wind_solar(n):
...
                         if carrier in index], 'lifetime']=costs.at[carrier_name,'lifetime']
 
+
+def create_network_topology(n, prefix):
+    """
+    create a network topology as the electric network,
+    returns a pandas dataframe with bus0, bus1 and length
+    """
+
+    topo = pd.DataFrame(columns=["bus0", "bus1", "length"])
```
Comment (fneum) [inline scripts/prepare_sector_network.py:61]: Not needed.

File: scripts/prepare_sector_network.py [hunk 5]
```diff
@@ -51,6 +51,33 @@ def add_lifetime_wind_solar(n):
         n.generators.loc[[index for index in n.generators.index.to_list()
                         if carrier in index], 'lifetime']=costs.at[carrier_name,'lifetime']
 
+
+def create_network_topology(n, prefix):
```
Comment (fneum) [inline scripts/prepare_sector_network.py:55]: Could work with defaults:
```suggestion
def create_network_topology(n, prefix="", connector = " -> "):
```

File: scripts/prepare_sector_network.py [hunk 6]
```diff
@@ -1353,6 +1381,16 @@ def add_biomass(network):
...
                                      index_col=0)
 
+     # costs for biomass transport
+    transport_costs = pd.read_csv(snakemake.input.biomass_transport,
+                                  index_col=0)
+
+    # potential per node distributed within country by population
+    biomass_pot_node = (biomass_potentials.loc[pop_layout.ct]
+                        .set_index(pop_layout.index)
+                        .mul(pop_layout.fraction, axis="index"))
```
Comment (fneum) [inline scripts/prepare_sector_network.py:1391]: This is where in the future we will distribute by ENSPRESO database

File: scripts/prepare_sector_network.py (hunk 1 above)
Comment (lisazeyen) [reply scripts/prepare_sector_network.py:2019]: yes, that would be more logical. Should we deal with it in a seperate pull-request and change the way the H2 network is implemented as well? I would suggest, that the functions (e.g. add_biomass_copperplate) only add the biomass node(s). We could then all the other functions e.g. add_industry(n) by filtering for all buses with the corresponding carrier ("solid biomass" or "H2"), e.g.
`biomass_nodes = n.buses[n.buses.carrier=="solid biomass]`
 and using them for adding additional components to the network.

File: scripts/prepare_sector_network.py (hunk 4 above)
Comment (lisazeyen) [reply scripts/prepare_sector_network.py:61]: True! Can be removed!

File: scripts/prepare_sector_network.py (hunk 5 above)
Comment (lisazeyen) [reply scripts/prepare_sector_network.py:55]: Yes, I agree!

**Comment by fneum** [comment]
- [ ] add `tabula-py` to environment (from conda-forge)

File: scripts/prepare_sector_network.py (hunk 1 above)
Comment (fneum) [reply scripts/prepare_sector_network.py:2188]: Yes, let's move that to a separate issue. I just updated everything to the revision #139

Comment (fneum) [reply scripts/prepare_sector_network.py:2188]: See proposal in #152

**Review by lisazeyen — COMMENTED** [review]
I only got one error when building the transport costs. I suggested a change and think this branch is ready to be merged.

**Pull request merged**